docs(codebuild): Add user guide for AWS CodeBuild stage (#1725)
spinnaker/spinnaker.github.io: setup/ci/codebuild.md

---
layout: single
title:  "AWS CodeBuild"
sidebar:
  nav: setup
---

{% include toc %}

Setting up [AWS CodeBuild](https://aws.amazon.com/codebuild/) as a Continuous Integration (CI)
system within Spinnaker allows you to:
 * trigger pipelines when an AWS CodeBuild build changes its phase or state
 * add an AWS CodeBuild stage to your pipeline

## Prerequisites

### AWS CodeBuild project

> You need to have an [AWS CodeBuild](https://aws.amazon.com/codebuild/) project

Reviewer: ```suggestion
You need to have an [AWS CodeBuild](https://aws.amazon.com/codebuild/) project.
```

        "codebuild:BatchGetBuilds"
      ],
      "Resource": "*"
    }
  ]
}
```
This role has to trust the Spinnaker auth role so that the managing account can assume this role to call CodeBuild.
You can edit your trust relationship to add the following policy.
```
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Effect": "Allow",
      "Principal": {
        "AWS": <Your Spinnaker Auth Role>
      },
      "Action": "sts:AssumeRole"
    }
  ]
}
```

## Configure Spinnaker to work with AWS CodeBuild

Use the following Halyard commands to create an AWS CodeBuild account, enable the AWS CodeBuild integration, and re-deploy Spinnaker:
```
hal config ci codebuild account add $ACCOUNT_NAME \
  --accountId $ACCOUNT_ID \
  --assume-role $ASSUME_ROLE \
  --region $REGION

hal config ci codebuild enable

hal deploy apply
```

## Configure an AWS CodeBuild stage

To run an AWS CodeBuild build as part of a Spinnaker pipeline:

1. Create a stage of type *AWS CodeBuild*.

2. Configure the stage by selecting the AWS CodeBuild account to use to run the build, and selecting the project name
from the dropdown list.

3. (Optional) In the *Source Configuration* section, you might 

>   - Select the source artifact to use as build source. If not specified, the source configured in project is used.

Reviewer: ```suggestion
  - Select the source artifact to use as the build source. If not specified, the source configured in the project is used.
```

  - Specify the [source version](https://docs.aws.amazon.com/codebuild/latest/APIReference/API_StartBuild.html#CodeBuild-StartBuild-request-sourceVersion)

>   of the build source.If not specified, the latest version is used.

Reviewer: ```suggestion
  of the build source. If not specified, the latest version is used.
```

>   - Specify the buildspec file. If left blank, the buildspec configured in project is used.

Reviewer: ```suggestion
  - Specify the buildspec file. If left blank, the buildspec configured in the project is used.
```

>   - Select secondary build sources. If not specified, the secondary sources configured in project is used

Reviewer: ```suggestion
  - Select secondary build sources. If not specified, the secondary sources configured in the project is used.
```

4. (Optional) In the *Environment Configuration* section, you might specify the image tag or image digest that identifies the Docker image

> to use. If not specified, the image configured in project is used. **Note:** As prerequisite, follow this

Reviewer: ```suggestion
to use. If not specified, the image configured in the project is used. **Note:** As a prerequisite, follow this
```

[user guide](https://docs.aws.amazon.com/codebuild/latest/userguide/sample-private-registry.html) to set up your registry credentials

> in project configuration.

Reviewer: ```suggestion
in the project configuration.
```

5. (Optional) In the *Advanced Configuration* section, you might specify environment variables to use in the build.

> Only environment variables in plain text are supported.

Reviewer: Might want to note here that Parameter Store / Secrets Manager environment variables can be configured in the project.

Author: will add

6. (Optional) In the *Produces Artifacts* section, you might supply any artifacts that you expect the build to create in order to
make these artifacts available to downstream stages. AWS CodeBuild supports creating S3 artifacts, which will be converted

> to Spinnaker artifacts and injected into the pipeline on completion of the build.

Reviewer: Are there any restrictions on the artifact packaging value in the project for this, like must be set to "zip"?  Can Spinnaker artifacts handle getting an S3 folder returned if "none" packaging is specified?

Author: Yes I think Spinnaker could only handle s3 object but not s3 folder
https://www.spinnaker.io/reference/artifacts/types/s3-object/

> While your build is executing, the stage details provide the following information

Reviewer: ```suggestion
While your build is running, the stage details provide the following information
```

* Current status of the build
* ARN of the build
* The link to view the build in the AWS CodeBuild Console
* The link to view the build logs in AWS CloudWatch Logs
* The link to view the build logs in S3

## Configure your pipeline trigger (Optional)

> Configure your pipeline to be triggered by AWS CodeBuild build phase or state changes:

Reviewer: Is this not required when the build is a stage in the pipeline?  Does Spinnaker poll for the completed build state while the stage is running a build?

Author: Spinnaker polls the build state by calling BatchGetBuilds every 10 seconds, so a codebuild trigger is not required.

1. Follow [instructions](https://docs.aws.amazon.com/codestar-notifications/latest/userguide/getting-started-build.html) to
set up a notification rule for your project.

2. Follow [instructions](/setup/triggers/amazon/) to create an Amazon Pub/Sub trigger.

>   - Skip the step to create SNS topic, as we will be using the SNS topic created in step 1.

Reviewer: ```suggestion
  - Skip the step to create SNS topic, as we will use the SNS topic created in step 1.
```

  - Create an SQS queue and subscribe to the SNS topic.

>   - Skip the step to create S3 bucket, as the notification will be sent from CodeBuild instead of S3.

Reviewer: ```suggestion
  - Skip the step to create an S3 bucket, as the notification will be sent from CodeBuild instead of from S3.
```

>   - Write Echo configuration and set up the Amazon Pub/Sub trigger in UI

Reviewer: ```suggestion
  - Write the Echo configuration and set up the Amazon Pub/Sub trigger in the Spinnaker UI.
```

> 3. Start a build in CodeBuild and your pipeline should be triggered.

Reviewer: Pipeline is triggered only when the build completes?

Author: Pipeline could be triggered by either build phase or build state changes

Author: or both

Reviewer: That's dependent on how they setup their notification rule, right?  It seems like it would only make sense if the pipeline triggered when the build completes so you can get the built artifacts later in the pipeline.  A suggestion of how to configure the notification rule to achieve that behavior would be useful here.

Author: Yes it depends on how they setup the notification rule. I will add steps here.

[user guide](https://docs.aws.amazon.com/codebuild/latest/userguide/sample-private-registry.html) to set up your registry credentials
in the project configuration.

5. (Optional) In the *Advanced Configuration* section, you might specify environment variables to use in the build.
Only environment variables in plain text are supported. Parameter Store / Secrets Manager environment variables
can be configured in the project.

6. (Optional) In the *Produces Artifacts* section, you might supply any artifacts that you expect the build to create in order to
make these artifacts available to downstream stages. AWS CodeBuild supports creating S3 artifacts in ZIP format, which will be converted
to Spinnaker artifacts and injected into the pipeline on completion of the build.

While your build is running, the stage details provide the following information
* Current status of the build
* ARN of the build
* The link to view the build in the AWS CodeBuild Console
* The link to view the build logs in AWS CloudWatch Logs
* The link to view the build logs in S3

## Configure your pipeline trigger (Optional)

> Configure your pipeline to be triggered when AWS CodeBuild build completes:

Reviewer: ```suggestion
Configure your pipeline to be triggered when an AWS CodeBuild build completes:
```

1. Follow [instructions](https://docs.aws.amazon.com/codestar-notifications/latest/userguide/getting-started-build.html) to
set up a notification rule for your project. To trigger the pipeline on completion of the build,

> select *Succeeded* and *Failed* in *Events that trigger notifications* section under *Build state*.

Reviewer: ```suggestion
select *Succeeded* and *Failed* in the *Events that trigger notifications* section under *Build state*.
```

2. Follow [instructions](/setup/triggers/amazon/) to create an Amazon Pub/Sub trigger.
  - Skip the step to create SNS topic, as we will use the SNS topic created in step 1.

>   - Create an SQS queue and subscribe to the SNS topic.

Reviewer: ```suggestion
  - Create an SQS queue and subscribe the queue to the SNS topic.
```
